x-pack/metricbeat/module/azure/monitor/monitor.go

```diff
@@ -0,0 +1,73 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package monitor
+
+import (
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/azure"
+	"github.com/pkg/errors"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("azure", "monitor", New, mb.DefaultMetricSet())
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	client *Client
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The azure monitor metricset is beta.")
+	var config azure.Config
+	err := base.Module().UnpackConfig(&config)
+	if err != nil {
+		return nil, errors.Wrap(err, "error unpack raw module config using UnpackConfig")
+	}
+	if len(config.Resources) == 0 {
+		return nil, errors.New("no metrics defined")
+	}
+	monitorClient, err := NewClient(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "error initializing the monitor client")
+	}
+	err = monitorClient.InitResources()
+	if err != nil {
+		return nil, errors.Wrap(err, "error initializing the monitor client")
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		client:        monitorClient,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	err := m.client.InitResources()
```

Comment:
I see this is also happening on line 47. Does it have to happen on each fetch or not?

Reply:
At the moment if users set a refreshInterval for the lists then we will retrieve the resources when the metricset starts and set an expiration period.
At each fetch we check if that timeline has expired, if so we will refresh the list, else we will skip and retrieve the metric values. 
The check happens inside InitResources func (specifically the expired func).